drop-struct-macro-derive/src/lib.rs

```diff
@@ -0,0 +1,107 @@
+extern crate proc_macro;
+use crate::proc_macro::TokenStream;
+use proc_macro2::{Ident, Span};
+use quote::{quote, ToTokens};
+use syn;
+
+/// A struct that contains the name of a struct field and the corresponding type
+#[derive(Debug)]
+struct FieldNameType {
+    field_name: String,
+    field_type: proc_macro2::TokenStream,
+}
+
+/// The actual code to free the *const pointers
+impl quote::ToTokens for FieldNameType {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        let field_type = &self.field_type;
+        // Convert field type to string for easy pattern matching
+        let field_type_string = self
+            .field_type
+            .clone()
+            .into_iter()
+            .map(|token| token.to_string())
+            .collect::<String>();
+        match &field_type_string[..] {
+            // Free string with `free_c_str`
+            "libc::c_char" => {
+                let field_name = Ident::new(&self.field_name, Span::call_site());
+                let gen = quote! {
+                    free_c_str(self.#field_name as *mut #field_type);
+                };
+                gen.to_tokens(tokens);
+            }
+            // Expect all others to be vectors
+            _ => {
```

Comment:
Can we explicitly capture `Vec`, and throw a nice error if it is not one?

Reply:
The comment might be misleading. There is no `Vec` here. There's fields ending with `_len` and `_ptr`. If that's not the case there's a panic (see next line), else we expect it to be a vector which is composed by those fields.

Reply:
Ah okay, that makes sense.